### ISSUE TITLE
podman rm: fails, and leaves everything hosed

### ISSUE THREAD

**Issue by edsantiago** [containers/podman#15367]
Another hard-to-isolate flake. I'm not convinced that it's truly podman-remote-only, because I've actually seen this one on my laptop. Two symptoms (possibly unrelated):
```
# podman-remote --url unix:/tmp/podman_tmp_loGI rm -t 0 -f cpcontainer
Error: cannot remove container 5d183a4ab677c5b65a78ebf87720153aee486145b24631f08edb7e20671498b7 as it could not be stopped: error sending SIGKILL to container 5d183a4ab677c5b65a78ebf87720153aee486145b24631f08edb7e20671498b7: %!w(<nil>)
[ rc=125 (** EXPECTED 0 **) ]
```
and
```
# podman-remote --url unix:/tmp/podman_tmp_fYnD pod rm -f -t 0 -a
Error: error freeing lock for container 38ecd608db0c5616563d5b268e85fea2fbada7a9b7d57eabf705c4f8f92aca92: no such file or directory
[ rc=125 (** EXPECTED 0 **) ]
```
With the second one, once it happens, the entire system becomes completely unusable.

Here's an [example of the cannot-remove one](https://api.cirrus-ci.com/v1/artifact/task/4550846103945216/html/sys-remote-fedora-36-aarch64-root-host.log.html#t--00140)

Here's an [example of the error-freeing one](https://api.cirrus-ci.com/v1/artifact/task/6115402563452928/html/sys-remote-fedora-36-aarch64-root-host.log.html#t--00266)

**Comment by edsantiago** [comment]
Here's the `error freeing` one, [sys remote f36 rootless](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/5555957253013504/html/sys-remote-fedora-36-rootless-host.log.html#t--00364), possibly connected to #11594

**Comment by edsantiago** [comment]
The `error freeing` one, [int podman f35 root](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/6191567470854144/html/int-podman-fedora-35-root-host.log.html#t--podman-pod-logs-with-container-names--1), May 27.

**Comment by rhatdan** [comment]
In runtime_ctr I see:
```
	// For pod removal, the container is already locked by the caller
	if !removePod {
		c.lock.Lock()
		defer c.lock.Unlock()
	}
...
	// Deallocate the container's lock
	if err := c.lock.Free(); err != nil {
		if cleanupErr == nil {
			cleanupErr = fmt.Errorf("error freeing lock for container %s: %w", c.ID(), err)
		} else {
			logrus.Errorf("Free container lock: %v", err)
		}
	}
```
Could this be a situation where the c.lock is never Locked?

**Comment by edsantiago** [comment]
Here's one with possibly useful context. [sys remote f36 rootless](https://api.cirrus-ci.com/v1/artifact/task/6408248566743040/html/sys-remote-fedora-36-rootless-host.log.html#t--00377):
```
$ podman-remote --url unix:/tmp/podman_tmp_r7Ne kube down /tmp/podman_bats.DD7B1h/test.yaml
Error: json: cannot unmarshal string into Go struct field PodRmReport.RmReport.Err of type error
```
Then, immediately after that, kablooie: `error freeing` on all subsequent tests.

**Comment by edsantiago** [comment]
Interesting. This one [sys remote ubuntu-2204 root](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/6169165353451520/html/sys-remote-ubuntu-2204-root-host.log.html#t--00270) goes into the `error freeing` death spiral, but the trigger seems to be the `unlinkat/EBUSY` flake (#11594):
```
# podman-remote --url unix:/tmp/podman_tmp_9X4G pod rm yky40edmyy
Error: error removing container c460399c59e569bd25127913cecc01c7de058a07c0aacde4bc9fa2ae2f2c3128 root filesystem: 1 error occurred:
	* unlinkat /var/lib/containers/storage/overlay/a6b8a4bc3e3ab118ed01e0b3e8d16f7b41a7b50176f6f9eb713c5baf9bbf47e0/merged: device or resource busy
[ rc=125 (** EXPECTED 0 **) ]
#/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
#| FAIL: exit code is 125; expected 0
#\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# podman-remote --url unix:/tmp/podman_tmp_9X4G pod rm -f -t 0 -a
Error: error freeing lock for container c460399c59e569bd25127913cecc01c7de058a07c0aacde4bc9fa2ae2f2c3128: no such file or directory
```

**Comment by apinter** [comment]
Not sure if at all related, but got a similar - likely storage related - issue as well on openSUSE MicroOS:

```bash
adathor@vegas:~> podman pod --log-level debug rm -i -f hedgedoc 
INFO[0000] podman filtering at log level debug          
DEBU[0000] Called rm.PersistentPreRunE(podman pod --log-level debug rm -i -f hedgedoc) 
DEBU[0000] Merged system config "/usr/share/containers/containers.conf" 
DEBU[0000] Using conmon: "/usr/bin/conmon"              
DEBU[0000] Initializing boltdb state at /home/podman_vol/home-vol/podman/containers/storage/libpod/bolt_state.db 
DEBU[0000] systemd-logind: Unknown object '/'.          
DEBU[0000] Using graph driver btrfs                     
DEBU[0000] Using graph root /home/podman_vol/home-vol/podman/containers/storage 
DEBU[0000] Using run root /run/user/1000/containers     
DEBU[0000] Using static dir /home/podman_vol/home-vol/podman/containers/storage/libpod 
DEBU[0000] Using tmp dir /run/user/1000/libpod/tmp      
DEBU[0000] Using volume path /home/podman_vol/home-vol/podman/containers/storage/volumes 
DEBU[0000] Set libpod namespace to ""                   
DEBU[0000] [graphdriver] trying provided driver "btrfs" 
DEBU[0000] Initializing event backend journald          
DEBU[0000] Configured OCI runtime crun initialization failed: no valid executable found for OCI runtime crun: invalid argument 
DEBU[0000] Configured OCI runtime runj initialization failed: no valid executable found for OCI runtime runj: invalid argument 
DEBU[0000] Configured OCI runtime kata initialization failed: no valid executable found for OCI runtime kata: invalid argument 
DEBU[0000] Configured OCI runtime runsc initialization failed: no valid executable found for OCI runtime runsc: invalid argument 
DEBU[0000] Configured OCI runtime krun initialization failed: no valid executable found for OCI runtime krun: invalid argument 
DEBU[0000] Using OCI runtime "/usr/bin/runc"            
INFO[0000] Setting parallel job count to 13             
DEBU[0000] Removing container 052bcd6e2d129159e7f8871310c06cc485a124bf52699db8ee41cb26e3c3e0fb 
DEBU[0000] Cleaning up container 052bcd6e2d129159e7f8871310c06cc485a124bf52699db8ee41cb26e3c3e0fb 
DEBU[0000] Network is already cleaned up, skipping...   
DEBU[0000] Container 052bcd6e2d129159e7f8871310c06cc485a124bf52699db8ee41cb26e3c3e0fb storage is already unmounted, skipping... 
DEBU[0000] Removing all exec sessions for container 052bcd6e2d129159e7f8871310c06cc485a124bf52699db8ee41cb26e3c3e0fb 
DEBU[0000] Container 052bcd6e2d129159e7f8871310c06cc485a124bf52699db8ee41cb26e3c3e0fb storage is already unmounted, skipping... 
INFO[0000] Storage for container 052bcd6e2d129159e7f8871310c06cc485a124bf52699db8ee41cb26e3c3e0fb already removed 
DEBU[0000] Removing container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b 
DEBU[0000] Cleaning up container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b 
DEBU[0000] Failed to reset unit file: "Unit 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b.service not loaded." 
DEBU[0000] Container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b storage is already unmounted, skipping... 
DEBU[0000] Removing all exec sessions for container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b 
DEBU[0000] Container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b storage is already unmounted, skipping... 
INFO[0000] Storage for container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b already removed 
ERRO[0000] Removing container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b from pod ad97dd32896ba8af8f4a76144f9a6f6c26cab1b352ef47a11e91af48cbbdf2d4: error freeing lock for container 200ddd92e43b4a2090682ab7534ec126fec80000dfe2f87041ed5236f3a8459b: no such file or directory 
```

```bash
adathor@vegas:~> podman info
host:
  arch: amd64
  buildahVersion: 1.27.0
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.2-1.1.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.2, commit: unknown'
  cpuUtilization:
    idlePercent: 93.86
    systemPercent: 2.51
    userPercent: 3.64
  cpus: 4
  distribution:
    distribution: '"opensuse-microos"'
    version: "20220822"
  eventLogger: journald
  hostname: vegas
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.19.2-1-default
  linkmode: dynamic
  logDriver: journald
  memFree: 16723730432
  memTotal: 20513906688
  networkBackend: cni
  ociRuntime:
    name: runc
    package: runc-1.1.3-2.1.x86_64
    path: /usr/bin/runc
    version: |-
      runc version 1.1.3
      commit: v1.1.3-0-ga916309fff0f
      spec: 1.0.2-dev
      go: go1.18.3
      libseccomp: 2.5.4
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.11-1.6.x86_64
    version: |-
      slirp4netns version 1.1.11
      commit: unknown
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.4
  swapFree: 0
  swapTotal: 0
  uptime: 0h 13m 0.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - registry.opensuse.org
  - registry.suse.com
  - docker.io
store:
  configFile: /home/adathor/.config/containers/storage.conf
  containerStore:
    number: 10
    paused: 0
    running: 7
    stopped: 3
  graphDriverName: btrfs
  graphOptions: {}
  graphRoot: /home/podman_vol/home-vol/podman/containers/storage
  graphRootAllocated: 10000830279680
  graphRootUsed: 247550713856
  graphStatus:
    Build Version: Btrfs v5.18.1
    Library Version: "102"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 28
  runRoot: /run/user/1000/containers
  volumePath: /home/podman_vol/home-vol/podman/containers/storage/volumes
version:
  APIVersion: 4.2.0
  Built: 1660176000
  BuiltTime: Thu Aug 11 07:00:00 2022
  GitCommit: ""
  GoVersion: go1.16.15
  Os: linux
  OsArch: linux/amd64
```

**Comment by edsantiago** [comment]
Here's another `unlinkat` (#11594) followed by systemsnafu: [sys remote ubuntu 2204 root](https://api.cirrus-ci.com/v1/artifact/task/6260854717939712/html/sys-remote-ubuntu-2204-root-host.log.html).

@containers/podman-maintainers since this is now in the wild, I'd say it's higher priority.

**Comment by edsantiago** [comment]
[f36 aarch64 root](https://api.cirrus-ci.com/v1/artifact/task/5872726341582848/html/sys-remote-fedora-36-aarch64-root-host.log.html)

**Comment by edsantiago** [comment]
No `podman rm` that I can see, but here we go again, [remote f36 root](https://api.cirrus-ci.com/v1/artifact/task/4874587107229696/html/sys-remote-fedora-36-root-host.log.html)

**Comment by edsantiago** [comment]
[remote f36-aarch64 root](https://api.cirrus-ci.com/v1/artifact/task/5872726341582848/html/sys-remote-fedora-36-aarch64-root-host.log.html#t--00267)

**Comment by edsantiago** [comment]
[remote f36 root](https://api.cirrus-ci.com/v1/artifact/task/5094388970815488/html/sys-remote-fedora-36-root-host.log.html#t--00265)

**Comment by edsantiago** [comment]
Here's another one with the `unmarshal` error happening first, then everything going bonkers. [remote ubuntu root](https://api.cirrus-ci.com/v1/artifact/task/5080405429714944/html/sys-remote-ubuntu-2204-root-host.log.html#t--00384)
```
# podman-remote --url unix:/tmp/podman_tmp_wtPS inspect --format {{.HostConfig.LogConfig.Type}} test_pod-test
journald
# podman-remote --url unix:/tmp/podman_tmp_wtPS kube down /tmp/podman_bats.LQsRLR/test.yaml
Error: json: cannot unmarshal string into Go struct field PodRmReport.RmReport.Err of type error
...
# podman-remote --url unix:/tmp/podman_tmp_wtPS pod rm -t 0 -f -a
Error: removing container d4d4bc7ee1505bd7738922c686fd35788c67e4c7a936e62044287dde0abaf6a1 root filesystem: 1 error occurred:
	* unlinkat /var/lib/containers/storage/overlay/c01ffe8ae42779fe4737be67fb350cabbe5f0c0a309ffbcd318fd584ff7af701/merged: device or resource busy
```

**Comment by edsantiago** [comment]
Interesting variation on [remote f36-aarch64 root](https://api.cirrus-ci.com/v1/artifact/task/5694475040194560/html/sys-remote-fedora-36-aarch64-root-host.log.html#t--00124): "cannot remove container" but subsequent tests actually pass:
```
# podman-remote --url unix:/tmp/podman_tmp_IPZS kill cpcontainer
cpcontainer
# podman-remote --url unix:/tmp/podman_tmp_IPZS rm -t 0 -f cpcontainer
Error: cannot remove container 380fcd0fd9264c2daf9157c7dfe07a8f7e31a84bd1c47e14a3ad01e76d4a7682 as it could not be stopped: sending SIGKILL to container 380fcd0fd9264c2daf9157c7dfe07a8f7e31a84bd1c47e14a3ad01e76d4a7682: container state improper
[ rc=2 (** EXPECTED 0 **) ]
```

**Comment by edsantiago** [comment]
Here's the unlinkat/ebusy one on [f36 remote rootless](https://api.cirrus-ci.com/v1/artifact/task/5022141530767360/html/sys-remote-fedora-36-rootless-host.log.html#t--00346)

OK, f2f over, we need serious attention on this one.

**Comment by edsantiago** [comment]
[f36 remote rootless](https://api.cirrus-ci.com/v1/artifact/task/5193909100871680/html/sys-remote-fedora-36-rootless-host.log.html#t--00390) again

**Comment by edsantiago** [comment]
Slightly different symptom here, also [f36 remote root](https://api.cirrus-ci.com/v1/artifact/task/6520997825216512/html/sys-remote-fedora-36-root-host.log.html#t--00266):
```
#  podman-remote run --rm --pod podOWvA8g8Ywr quay.io/libpod/testimage:20220615 ls /dev/shm/test1
/dev/shm/test1
time="2022-09-30T07:21:59-05:00" level=error msg="Removing container SHA: 
   container \"SHA\" is mounted and cannot be removed without using force: container state improper"
```
It's followed by the usual cascade of everything-is-broken-forever failures, so I'm treating it as the same bug.

**Comment by edsantiago** [comment]
[f36 remote aarch64 root](https://api.cirrus-ci.com/v1/artifact/task/5818176876314624/html/sys-remote-fedora-36-aarch64-root-host.log.html)

**Comment by edsantiago** [comment]
Another new failure mode. Seen in [remote f36-aarch64 root](https://api.cirrus-ci.com/v1/artifact/task/5818176876314624/html/sys-remote-fedora-36-aarch64-root-host.log.html#t--00360):
```
# podman-remote rm -t 0 --all --force --ignore
# podman-remote ps --all --external --format {{.ID}} {{.Names}}
# podman-remote images --all --format {{.Repository}}:{{.Tag}} {{.ID}}
Error: retrieving parent of image "SHA1": 
     you may need to remove the image to resolve the error: 
     unlinkat /var/lib/containers/storage/overlay/SHA2/merged: device or resource busy
```
In this one, there's no `podman rm` that I can see (no failing one, I mean). The failure happens on `podman images`. (And yes, as with all other instances, causes every subsequent test to die).

Can someone please treat this one as serious?

**Comment by vrothberg** [comment]
The symptoms look similar to https://github.com/containers/podman/issues/11594#issuecomment-1263642097, don't they?

**Comment by edsantiago** [comment]
Sometimes, but not enough for me to combine the two issues. #11594 is a one-shot: it happens, and life goes on. This issue is a catastrophe: once it happens, podman never works again.

**Comment by edsantiago** [comment]
Potentially interesting correlation: #16154 (unmarshal blah blah, podman-remote only) seems to be highly correlated with this one. My gut feeling is that "unmarshal" is its own bug; that what's happening is that THIS issue (the forever-hose) is triggering, and sending an error that podman-remote cannot deal with. So the underlying error is actually this one. But please take this with a grain of salt.

**Comment by edsantiago** [comment]
This one is really hitting us hard. Last ten days:

* fedora-36 : sys remote fedora-36 root host [remote]
  * PR #16223
    * [10-19 09:20](https://api.cirrus-ci.com/v1/artifact/task/5066313759457280/html/sys-remote-fedora-36-root-host.log.html#t--00271)
  * PR #16221
    * [10-18 18:10](https://api.cirrus-ci.com/v1/artifact/task/5884007350009856/html/sys-remote-fedora-36-root-host.log.html#t--00361)
  * PR #16211
    * [10-19 07:00](https://api.cirrus-ci.com/v1/artifact/task/4867011506339840/html/sys-remote-fedora-36-root-host.log.html#t--00272)
  * PR #16207
    * [10-17 16:27](https://api.cirrus-ci.com/v1/artifact/task/6627360156942336/html/sys-remote-fedora-36-root-host.log.html#t--00275)
  * PR #16177
    * [10-14 09:03](https://api.cirrus-ci.com/v1/artifact/task/5295287592812544/html/sys-remote-fedora-36-root-host.log.html#t--00270)
  * PR #16151
    * [10-13 08:10](https://api.cirrus-ci.com/v1/artifact/task/5263892455620608/html/sys-remote-fedora-36-root-host.log.html#t--00272)
  * PR #16011
    * [10-17 04:55](https://api.cirrus-ci.com/v1/artifact/task/5579225229950976/html/sys-remote-fedora-36-root-host.log.html#t--00267)
* fedora-36 : sys remote fedora-36 rootless host [remote]
  * PR #13808
    * [10-14 22:48](https://api.cirrus-ci.com/v1/artifact/task/5121289340846080/html/sys-remote-fedora-36-rootless-host.log.html#t--00349)
* fedora-36-aarch64 : sys remote fedora-36-aarch64 root host [remote]
  * PR #16127
    * [10-12 09:58](https://api.cirrus-ci.com/v1/artifact/task/5054141855694848/html/sys-remote-fedora-36-aarch64-root-host.log.html#t--00272)

**Comment by edsantiago** [comment]
Another case where the first failure presents with "blah blah is mounted", then cascades from there into disaster:
```
not ok 361 podman run with slirp4ns adds correct dns address to resolv.conf
...
# podman-remote run --rm  \
      --network slirp4netns:cidr=172.24.205.0/24 \
      quay.io/libpod/testimage:20221018 \
      grep 172.24.205 /etc/resolv.conf
nameserver 172.24.205.3
time="2022-10-25T13:47:23Z" level=error msg="Removing container SHA: 
         container \"SHA\" is mounted and cannot be removed without using force: container state improper"
...
[all subsequent tests fail]
```
[ubuntu 2204 root remote](https://api.cirrus-ci.com/v1/artifact/task/5336868848926720/html/sys-remote-ubuntu-2204-root-host.log.html#t--00361)

Can we please, please get some attention on this? It's a bad one, flaking at least once per day.

**Comment by edsantiago** [comment]
* fedora-36 : sys remote fedora-36 root host [remote]
  * PR #16459
    * [11-09 14:51](https://api.cirrus-ci.com/v1/artifact/task/4900286095949824/html/sys-remote-fedora-36-root-host.log.html#t--00271)
* ubuntu-2204 : sys podman ubuntu-2204 root host
  * PR #16465
    * [11-09 17:04](https://api.cirrus-ci.com/v1/artifact/task/5121353766928384/html/sys-podman-ubuntu-2204-root-host.log.html#t--00393)
* ubuntu-2204 : sys remote ubuntu-2204 root host [remote]
  * PR #16311
    * [10-27 15:03](https://api.cirrus-ci.com/v1/artifact/task/6399263438012416/html/sys-remote-ubuntu-2204-root-host.log.html#t--00361)

**Comment by vrothberg** [comment]
@containers/podman-maintainers PTAL

**Comment by rhatdan** [comment]
The issue is the 		timesMounted, err := r.store.Mounted(ctr.ID) call below could be getting screwed up or have a race condition.
```
	if !force {
		timesMounted, err := r.store.Mounted(ctr.ID)
		if err != nil {
			if errors.Is(err, storage.ErrContainerUnknown) {
				// Container was removed from under us.
				// It's gone, so don't bother erroring.
				logrus.Infof("Storage for container %s already removed", ctr.ID)
				return nil
			}
			logrus.Warnf("Checking if container %q is mounted, attempting to delete: %v", idOrName, err)
		}
		if timesMounted > 0 {
			return fmt.Errorf("container %q is mounted and cannot be removed without using force: %w", idOrName, define.ErrCtrStateInvalid)
		}
```

It is possible to loose an unmount on a killed process. This couner is attempting to keep track of the number of 

podman mount XYZ
podman mount XYZ
podman run XYZ ...

timesMounted would show 3.  If for some reason we had a race where someone else was executing a mount while executing this code, it could cause issues. Or if a podman container cleanup failed and the number of mounts was not decremented.

**Comment by edsantiago** [comment]
Now seen in [f37 gating tests, too](https://artifacts.dev.testing-farm.io/2fb0707b-d2f9-43f5-9ee0-cfec3e4acbbd/work-tests.ymlM4I85_/tests-GyiFPE/test.podman-remote-rootless.bats.log)

**Comment by Johnpc123** [comment]
...also can't remove a pod after, being lazy, and attempting rm -f on a running pod.
  
`"Error: error freeing lock for container 95de07970b6cd967f1587294ffe05db767cd2f20bfd061633b010bb3654e2d73: no such file or directory"
`
How can I help?

**Comment by vrothberg** [comment]
@mtrmac I think your recent c/storage fixes may have resolved this issue.  It seems to happen only with remote which somehow made me think about your in-process fixes.

WDYT?

**Comment by mtrmac** [comment]
This is quite long, and might well mix different root causes:
- `error sending SIGKILL to container 5d183a4ab677c5b65a78ebf87720153aee486145b24631f08edb7e20671498b7: %!w(<nil>)`:
    - the `nil` part was fixed in 6bf8670b694278e5e53655ad40605d099f7ae02b
    - The error itself seems unlikely to be directly c/storage-related ; c/storage does not record PIDs or process state.
- `freeing lock for container`: These are `libpod/lock.Manager` locks
- `unlinkat /var/lib/containers/storage/overlay/…/merged: device or resource busy`: something is missing an unmount, or otherwise mismanaging recorded state vs. running processes??
- `podman-remote images … retrieving parent of image … unlinkat /var/lib/containers/storage/overlay/SHA2/merged: device or resource busy`: Why is obtaining a parent (pure metadata) doing any overlay operations at all?
- “The issue is the timesMounted, err := r.store.Mounted(ctr.ID) call below could be getting screwed up”: That API is _fundamentally_ racy; as far as c/storage cares, the caller is not holding any lock and the data can be out-of date already at the moment it is returned by c/storage. Compare https://github.com/containers/storage/pull/1346/commits/cbc2ca843648d1ed547f9921cf9bbc6eaa2de5c7 . It may well be the case that Podman is doing more synchronization, I wouldn’t know.

In summary nothing of the above is obviously related to the c/storage work I was doing. The _container_ state cleanup (unlike layer state cleanup) is very minimal in c/storage.

My hunch would be to follow the `define.ErrCtrStateInvalid` inconsistencies. Not that I know how.

**Comment by vrothberg** [comment]
Thanks for taking a look, @mtrmac !

**Comment by vrothberg** [comment]
@edsantiago fresh occurrences of the flake since https://github.com/containers/podman/pull/16886?

**Comment by edsantiago** [comment]
@vrothberg it's hard to say: there have been a _lot_ of flakes in the past few weeks, even despite a low number of CI runs, and it'll take me half a day to evaluate/categorize them all. From a quick overview, though, I don't see the classing "everything hosed" signature. I'll keep my eyes open for it.

PS isn't today a holiday for you?

**Comment by vrothberg** [comment]
Thanks, Ed!

> PS isn't today a holiday for you?

Not in France.  But in the US, so please don't let me bother you today.  There's plenty of time for that this year :)

**Comment by edsantiago** [comment]
Here goes. I'm not 100% sure that these are the same bug, because they don't exhibit the "everything-hosed" symptom, but they _do_ look like the same "podman rm fails" symptom. Is it possible that #16886 fixed the everything-hosed bug, but we're still seeing the "could not remove container, could not be stopped" bug?


* fedora-36 : sys podman fedora-36 rootless host
  * PR #16822
    * [12-13 02:48](https://api.cirrus-ci.com/v1/artifact/task/5773202765381632/html/sys-podman-fedora-36-rootless-host.log.html#t--00312)
* fedora-37-aarch64 : sys remote fedora-37-aarch64 root host [remote]
  * PR #16970
    * [01-02 08:09](https://api.cirrus-ci.com/v1/artifact/task/5759446589112320/html/sys-remote-fedora-37-aarch64-root-host.log.html#t--00367)
  * PR #16893
    * [12-20 10:09](https://api.cirrus-ci.com/v1/artifact/task/5853022316658688/html/sys-remote-fedora-37-aarch64-root-host.log.html#t--00361)
  * PR #16794
    * [12-16 23:00](https://api.cirrus-ci.com/v1/artifact/task/6436209417781248/html/sys-remote-fedora-37-aarch64-root-host.log.html#t--00355)
